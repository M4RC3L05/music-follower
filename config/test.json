{
  "apps": {
    "jobs": {
      "sync-releases": {
        "max-release-time": "1 year"
      }
    }
  },
  "database": {
    "path": ":memory:"
  },
  "logger": {
    "level": 4,
    "color": false
  }
}
