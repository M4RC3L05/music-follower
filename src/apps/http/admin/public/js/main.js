import { RouterProvider } from "react-router-dom";
import { StrictMode } from "react";
import { createRoot } from "react-dom/client";

import html from "./common/html.js";
import router from "./pages/mod.js";

createRoot(globalThis.document.querySelector("#app")).render(html`
  <${StrictMode}>
    <${RouterProvider} router=${router} />
  <//>
`);
