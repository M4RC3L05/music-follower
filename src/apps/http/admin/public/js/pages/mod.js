import { Route, createHashRouter, createRoutesFromElements, redirect } from "react-router-dom";

import Layout from "../components/layout.js";
import html from "../common/html.js";

const router = createHashRouter(
  createRoutesFromElements(html`
    <${Route} path="/" element=${html`<${Layout} />`}>
      <${Route} index lazy=${() => import("./index-page.js")} />
      <${Route} path="artists" lazy=${() => import("./artists-page.js")}>
        <${Route} path="remote" lazy=${() => import("./artists-remote-page.js")} />
        <${Route}
          path="reset"
          loader=${() => {
            redirect("/artists", { replace: true });
            return null;
          }}
        />
      <//>
      <${Route}
        path="releases"
        lazy=${() => import("./releases-page.js")}
        shouldRevalidate=${({ currentUrl, nextUrl }) =>
          !currentUrl.searchParams.has("selectedRelease") && !nextUrl.searchParams.has("selectedRelease")}
      />
    <//>
  `),
);

export default router;
