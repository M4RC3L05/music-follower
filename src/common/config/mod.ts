import vine from "@vinejs/vine";

const configSchema = vine.object({
  apps: vine.object({
    web: vine.object({
      host: vine.string(),
      port: vine.number(),
      basicAuth: vine.object({
        username: vine.string(),
        password: vine.string(),
      }),
      feed: vine.object({
        maxReleases: vine.number(),
      }),
    }),
    jobs: vine.object({
      syncReleases: vine.object({
        maxReleaseTime: vine.string(),
      }),
    }),
  }),
  database: vine.object({
    path: vine.string(),
  }),
  remote: vine.object({
    itunes: vine.object({
      search: vine.object({
        url: vine.string(),
        searchArtists: vine.object({
          limit: vine.number(),
        }),
      }),
      lookup: vine.object({
        url: vine.string(),
        getLatestsArtistMusicReleases: vine.object({
          limit: vine.number(),
        }),
        getLatestsArtistAlbumReleases: vine.object({
          limit: vine.number(),
        }),
      }),
    }),
  }),
  csrf: vine.object({
    origin: vine.string(),
  }),
  media: vine.object({
    remoteArtistsPlaceholderImage: vine.string(),
  }),
});
const configSchemaValidator = vine.compile(configSchema);

const config = await configSchemaValidator.validate({
  apps: {
    web: {
      host: Deno.env.get("APPS_WEB_HOST") ?? "127.0.0.1",
      port: Deno.env.get("APPS_WEB_PORT") ?? 4321,
      basicAuth: {
        username: Deno.env.get("APPS_WEB_BASIC_AUTH_USERNAME") ?? "foo",
        password: Deno.env.get("APPS_WEB_BASIC_AUTH_PASSWORD") ?? "bar",
      },
      feed: {
        maxReleases: Deno.env.get("APPS_WEB_FEED_MAX_RELEASES") ?? 50,
      },
    },
    jobs: {
      syncReleases: {
        maxReleaseTime: Deno.env.get("JOBS_SYNC_RELEASES_MAX_RELEASE_TIME") ??
          "1 year",
      },
    },
  },
  database: {
    path: Deno.env.get("DATABASE_PATH") ?? "./data/app.db",
  },
  remote: {
    itunes: {
      search: {
        url: Deno.env.get("REMOTE_ITUNES_SEARCH_URL") ??
          "https://itunes.apple.com/search",
        searchArtists: {
          limit: Deno.env.get("REMOTE_ITUNES_SEARCH_SEARCH_ARTISTS_LIMIT") ?? 8,
        },
      },
      lookup: {
        url: Deno.env.get("REMOTE_ITUNES_LOOKUP_URL") ??
          "https://itunes.apple.com/lookup",
        getLatestsArtistMusicReleases: {
          limit: Deno.env.get(
            "REMOTE_ITUNES_LOOKUP_URL_GET_LATESTS_ARTIST_MUSIC_RELEASES_LIMIT",
          ) ?? 70,
        },
        getLatestsArtistAlbumReleases: {
          limit: Deno.env.get(
            "REMOTE_ITUNES_LOOKUP_URL_GET_LATESTS_ARTIST_ALBUM_RELEASES_LIMIT",
          ) ?? 50,
        },
      },
    },
  },
  csrf: {
    origin: Deno.env.get("CSRF_ORIGIN") ?? "\\.*",
  },
  media: {
    remoteArtistsPlaceholderImage: Deno.env.get("MEDIA_PLACEHOLDER_IMAGE") ??
      "/public/images/remote-artist-image-default.svg",
  },
});

export default config;
