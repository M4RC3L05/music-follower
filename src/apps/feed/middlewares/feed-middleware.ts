import { sql } from "@m4rc3l05/sqlite-tag";
import config from "config";
import { Feed } from "feed";
import type { Context } from "@hono/hono";
import type { Release } from "#src/database/types/mod.ts";
import type {
  ItunesLookupAlbumModel,
  ItunesLookupSongModel,
} from "#src/remote/mod.ts";

export const feedMiddleware = (c: Context) => {
  const data = c.get("database").all<Release>(sql`
    select *
    from releases
    where "releasedAt" <= strftime('%Y-%m-%dT%H:%M:%fZ' , 'now')
      and not exists(
        select true from json_each(hidden)
        where json_each.value = 'feed'
      )
    order by "feedAt" desc
    limit ${config.get<number>("apps.feed.maxReleases")};
  `);

  const feed = new Feed({
    title: "Music releases",
    description: "Get the latest music releases from artist you follow",
    id: "music_follower",
    language: "en",
    copyright: "Music Follower",
    updated: new Date(),
    generator: "Music Follower",
  });

  for (const release of data) {
    const parsedMetadata = JSON.parse(release.metadata) as
      | ItunesLookupAlbumModel
      | ItunesLookupSongModel;

    feed.addItem({
      date: new Date(release.feedAt),
      description: `
      <img src="${release.coverUrl}" />
      <h1>${release.name}</h1><p>${release.artistName}</p>
      <p>${new Date(release.releasedAt).getFullYear()}</p>
      `.trim(),
      title: `${release.name} by ${release.artistName}`,
      id: String(release.id),
      link: release.type === "collection"
        ? parsedMetadata.collectionViewUrl ?? release.coverUrl
        : release.type === "track"
        ? (parsedMetadata as ItunesLookupSongModel).trackViewUrl ??
          parsedMetadata.collectionViewUrl ??
          release.coverUrl
        : release.coverUrl,
    });
  }

  const accepts = c.req.header("accept");

  if (
    accepts?.includes("application/rss+xml") ??
      accepts?.includes("application/xml")
  ) {
    return c.body(feed.rss2(), 200, {
      "content-type": accepts?.includes("application/xml")
        ? "application/xml"
        : "application/rss+xml",
    });
  }

  if (accepts?.includes("application/atom+xml")) {
    return c.body(feed.atom1(), 200, {
      "content-type": "application/atom+xml",
    });
  }

  if (accepts?.includes("application/json")) {
    return c.body(feed.json1(), 200, { "content-type": "application/json" });
  }

  return c.body(feed.rss2(), 200, { "content-type": "application/xml" });
};
