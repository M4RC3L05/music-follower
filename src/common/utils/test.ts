import type { Hono } from "@hono/hono";
import { encodeBase64 } from "@std/encoding";
import { JSDOM } from "jsdom";
import config from "#src/common/config/mod.ts";
import {
  assertEquals,
  assertExists,
  AssertionError,
  assertStringIncludes,
  assertThrows,
} from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import toDiffableHtml from "diffable-html";

const basicAuth = config.apps.web.basicAuth;

export const requestAuth = (
  app: Hono,
  ...args: Parameters<Hono["request"]>
) => {
  const basicAuthHeaderValue = `Basic ${
    encodeBase64(`${basicAuth.username}:${basicAuth.password}`)
  }`;

  args[1] ??= {};

  const newHeaders: Record<string, string> = (
    args[1].headers instanceof Headers
      ? Object.fromEntries(args[1].headers.entries())
      : (args[1].headers ?? {})
  ) as Record<string, string>;

  newHeaders["authorization"] = basicAuthHeaderValue;

  args[1].headers = newHeaders;

  return app.request(...args);
};

export const getDom = async (response: Response) => {
  return new JSDOM(await response.text());
};

export const prettyFormatHTML = (input: string) => {
  return toDiffableHtml(input);
};

const getAssertSeeContent = (ele: Element, textOnly?: boolean) => {
  return textOnly ? ele.textContent : ele.outerHTML;
};

export const assertSeeTextInOrder = (
  dom: JSDOM,
  what: string[],
  where?: string,
) => {
  const contentStr = getAssertSeeContent(
    where
      ? dom.window.document.querySelector(where)!
      : dom.window.document as unknown as Element,
    true,
  );

  if (!contentStr) {
    throw new AssertionError("Could not get text content");
  }

  let lastIndex = 0;

  for (const item of what) {
    const contentSliced = contentStr.slice(lastIndex);
    lastIndex = contentSliced.indexOf(item);

    if (lastIndex === -1) {
      throw new AssertionError(
        `Could not see text content matching in order:\n\n${item}\n\nin:\n\n${contentSliced}`,
      );
    }
  }
};

export const assertSeeText = (dom: JSDOM, what: string, where?: string) => {
  const contentStr = getAssertSeeContent(
    where
      ? dom.window.document.querySelector(where)!
      : dom.window.document as unknown as Element,
    true,
  );

  if (!contentStr) {
    throw new AssertionError("Could not get text content");
  }

  assertStringIncludes(
    contentStr,
    what,
    `Could not see text content matching:\n\n${what}\n\nin:\n${contentStr}`,
  );
};

export const assertNotSeeText = (dom: JSDOM, what: string, where?: string) => {
  const contentStr = getAssertSeeContent(
    where
      ? dom.window.document.querySelector(where)!
      : dom.window.document as unknown as Element,
    true,
  );

  assertThrows(
    () => assertSeeText(dom, what, where),
    `Could see text content matching:\n\n${what}\n\nin:\n\n${contentStr}`,
  );
};

export const assertSeeInOrder = (
  dom: JSDOM,
  what: string[],
  where?: string,
) => {
  const contentStr = where
    ? getAssertSeeContent(dom.window.document.querySelector(where)!, false)
    : dom.serialize();

  if (!contentStr) {
    throw new AssertionError("Could not get text content");
  }

  let lastIndex = 0;

  for (const item of what) {
    const contentSliced = contentStr.slice(lastIndex);
    lastIndex = contentSliced.indexOf(item);

    if (lastIndex === -1) {
      throw new AssertionError(
        `Could see text matching in order:\n\n${item}\n\nin:\n\n${
          prettyFormatHTML(contentSliced)
        }`,
      );
    }
  }
};

export const assertSee = (dom: JSDOM, what: string, where?: string) => {
  const contentStr = where
    ? getAssertSeeContent(dom.window.document.querySelector(where)!, false)
    : dom.serialize();

  if (!contentStr) {
    throw new AssertionError("Could not get text content");
  }

  assertEquals(
    contentStr.includes(what),
    true,
    `Could not see text matching:\n\n${what}\n\nin:\n${
      prettyFormatHTML(contentStr)
    }`,
  );
};

export const assertNotSee = (dom: JSDOM, what: string, where?: string) => {
  const contentStr = where
    ? getAssertSeeContent(dom.window.document.querySelector(where)!, false)!
    : dom.serialize();

  assertThrows(
    () => assertSee(dom, what, where),
    `Could see an text matching:\n\n${what}\n\nin:\n${
      prettyFormatHTML(contentStr)
    }`,
  );
};

export const assertNodeNotExists = (
  dom: JSDOM,
  what: string,
  where?: string,
) => {
  const container = where
    ? dom.window.document.querySelector(where)
    : dom.window.document;

  if (!container) {
    throw new AssertionError("Could not get container element");
  }

  assertExists(
    container.querySelector(what) ? undefined : true,
    `Could find an element matching:\n\n${what}\n\nin\n${
      prettyFormatHTML(
        container === dom.window.document
          ? dom.serialize()
          : (container as Element).innerHTML,
      )
    }"`,
  );
};

export const assertNodeExists = (dom: JSDOM, what: string, where?: string) => {
  const container = where
    ? dom.window.document.querySelector(where)
    : dom.window.document;

  if (!container) {
    throw new AssertionError("Could not get container element");
  }

  assertExists(
    container.querySelector(what),
    `Could not find an element matching:\n\n${what}\n\nin\n${
      prettyFormatHTML(
        container === dom.window.document
          ? dom.serialize()
          : (container as Element).innerHTML,
      )
    }"`,
  );
};

export const expectHtmlSnapshot = async (
  t: Deno.TestContext,
  dom: JSDOM,
  where?: string,
) => {
  const content = where
    ? dom.window.document.querySelector(where)!.outerHTML
    : dom.serialize();
  await assertSnapshot(
    t,
    prettyFormatHTML(content),
  );
};
