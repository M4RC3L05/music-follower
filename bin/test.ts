import { run } from "#src/commands/test.js";

await run();
