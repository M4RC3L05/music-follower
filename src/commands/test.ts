import process from "node:process";
import { run as testRunner } from "node:test";

import { promise as glob } from "glob-promise";

export const run = async () => {
  testRunner({
    concurrency: true,
    files: await glob("./src/**/*.test.ts"),
  }).pipe(process.stdout);
};
