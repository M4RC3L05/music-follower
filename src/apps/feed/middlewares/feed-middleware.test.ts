/* eslint-disable import/no-named-as-default-member */
/* eslint-disable @typescript-eslint/ban-ts-comment */
/* eslint-disable @typescript-eslint/no-unsafe-call */
import assert from "node:assert";
import { afterEach, before, beforeEach, describe, it } from "node:test";

import config from "config";
import sinon from "sinon";

import { feedMiddleware } from "#src/apps/feed/middlewares/feed-middleware.js";
import { releaseFixtures } from "#src/utils/tests/fixtures/index.js";
import { databaseHooks } from "#src/utils/tests/hooks/index.js";

describe("feedMiddleware()", () => {
  before(async () => {
    await databaseHooks.migrate();
  });

  beforeEach(() => {
    sinon.useFakeTimers(0);
  });

  afterEach(() => {
    sinon.restore();
    sinon.reset();
  });

  it("should fallback to rss if no accepts header is sent", async () => {
    const ctx = { request: { accepts: () => false } };

    feedMiddleware(ctx as any);

    assert.strict.equal((ctx as any).type, "application/xml");

    // @ts-expect-error
    await assert.strict.snapshot(
      (ctx as any).body,
      "feedMiddleware() - should fallback to rss if no accepts header is sent",
    );
  });

  it("should send rss feed if it accepts `application/xml`", async () => {
    const ctx = { request: { accepts: (x: any) => x === "application/xml" } };

    feedMiddleware(ctx as any);

    assert.strict.equal((ctx as any).type, "application/xml");

    // @ts-expect-error
    await assert.strict.snapshot(
      (ctx as any).body,
      "feedMiddleware() - should send rss feed if it accepts `application/xml`",
    );
  });

  it("should send rss feed if it accepts `application/rss+xml`", async () => {
    const ctx = { request: { accepts: (x: any) => x === "application/rss+xml" } };

    feedMiddleware(ctx as any);

    assert.strict.equal((ctx as any).type, "application/rss+xml");

    // @ts-expect-error
    await assert.strict.snapshot(
      (ctx as any).body,
      "feedMiddleware() - should send rss feed if it accepts `application/rss+xml`",
    );
  });

  it("should send atom feed if it accepts `application/atom+xml`", async () => {
    const ctx = { request: { accepts: (x: any) => x === "application/atom+xml" } };

    feedMiddleware(ctx as any);

    assert.strict.equal((ctx as any).type, "application/atom+xml");

    // @ts-expect-error
    await assert.strict.snapshot(
      (ctx as any).body,
      "feedMiddleware() - should send atom feed if it accepts `application/atom+xml`",
    );
  });

  it("should send json feed if it accepts `application/json`", async () => {
    const ctx = { request: { accepts: (x: any) => x === "application/json" } };

    feedMiddleware(ctx as any);

    assert.strict.equal((ctx as any).type, "application/json");

    // @ts-expect-error
    await assert.strict.snapshot(
      (ctx as any).body,
      "feedMiddleware() - should send json feed if it accepts `application/json`",
    );
  });

  for (const ct of ["application/rss+xml", "application/atom+xml", "application/json"]) {
    it(`should send feed with elements for "${ct}"`, async () => {
      const ctx = { request: { accepts: (x: any) => x === ct } };

      databaseHooks.cleanup();
      releaseFixtures.loadRelease({ id: 1, feedAt: new Date(0) });
      releaseFixtures.loadRelease({ id: 2, feedAt: new Date(500) });
      releaseFixtures.loadRelease({ id: 3, feedAt: new Date(1000) });
      releaseFixtures.loadRelease({ id: 4, feedAt: new Date(1500) });

      sinon.stub(config, "get").returns(3);

      feedMiddleware(ctx as any);

      assert.strict.equal((ctx as any).type, ct);

      // @ts-expect-error
      await assert.strict.snapshot((ctx as any).body, `feedMiddleware() - should send feed with elements for "${ct}"`);
    });
  }

  it("should detect correctly the feed item link", async () => {
    const ctx = { request: { accepts: (x: any) => false } };

    databaseHooks.cleanup();
    releaseFixtures.loadRelease({
      id: 1,
      type: "track",
      metadata: { trackViewUrl: "foo", collectionViewUrl: "bar" },
      feedAt: new Date(0),
      coverUrl: "bix",
    });
    releaseFixtures.loadRelease({
      id: 2,
      type: "track",
      metadata: { trackViewUrl: undefined, collectionViewUrl: "bar" },
      feedAt: new Date(1),
      coverUrl: "bix",
    });
    releaseFixtures.loadRelease({
      id: 3,
      type: "track",
      metadata: { trackViewUrl: undefined, collectionViewUrl: undefined },
      feedAt: new Date(2),
      coverUrl: "bix",
    });
    releaseFixtures.loadRelease({
      id: 2,
      type: "collection",
      metadata: { collectionViewUrl: "foo" },
      feedAt: new Date(3),
      coverUrl: "bux",
    });
    releaseFixtures.loadRelease({
      id: 3,
      type: "collection",
      metadata: { collectionViewUrl: undefined },
      feedAt: new Date(4),
      coverUrl: "bux",
    });
    releaseFixtures.loadRelease({
      id: 1,
      type: "bix",
      metadata: { collectionViewUrl: undefined },
      feedAt: new Date(5),
      coverUrl: "bux",
    });

    sinon.stub(config, "get").returns(10);

    feedMiddleware(ctx as any);

    assert.strict.equal((ctx as any).type, "application/xml");

    // @ts-expect-error
    await assert.strict.snapshot((ctx as any).body, "feedMiddleware() - should detect correctly the feed item link");
  });
});
