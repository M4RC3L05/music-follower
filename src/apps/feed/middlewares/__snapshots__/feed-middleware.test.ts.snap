export const snapshot = {};

snapshot[`global > GET / > should send body as "application/rss+xml" if accepts is "application/rss+xml" 1`] = `
'<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0">
    <channel>
        <title>Music releases</title>
        <link>undefined</link>
        <description>Get the latest music releases from artist you follow</description>
        <lastBuildDate>Thu, 01 Jan 1970 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>Music Follower</generator>
        <language>en</language>
        <copyright>Music Follower</copyright>
        <item>
            <title><![CDATA[foo by bar]]></title>
            <link>http://example.com</link>
            <guid isPermaLink="false">1</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src="http://example.com" />
          <h1>foo</h1><p>bar</p>
          <p>1970</p>]]></description>
        </item>
    </channel>
</rss>'
`;

snapshot[`global > GET / > should send body as "application/xml" if accepts is "application/xml" 1`] = `
'<?xml version="1.0" encoding="utf-8"?>
<rss version="2.0">
    <channel>
        <title>Music releases</title>
        <link>undefined</link>
        <description>Get the latest music releases from artist you follow</description>
        <lastBuildDate>Thu, 01 Jan 1970 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>Music Follower</generator>
        <language>en</language>
        <copyright>Music Follower</copyright>
        <item>
            <title><![CDATA[foo by bar]]></title>
            <link>http://example.com</link>
            <guid isPermaLink="false">1</guid>
            <pubDate>Thu, 01 Jan 1970 00:00:00 GMT</pubDate>
            <description><![CDATA[<img src="http://example.com" />
          <h1>foo</h1><p>bar</p>
          <p>1970</p>]]></description>
        </item>
    </channel>
</rss>'
`;

snapshot[`global > GET / > should send body as "application/atom+xml" if accepts is "application/atom+xml" 1`] = `
'<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">
    <id>music_follower</id>
    <title>Music releases</title>
    <updated>1970-01-01T00:00:00.000Z</updated>
    <generator>Music Follower</generator>
    <subtitle>Get the latest music releases from artist you follow</subtitle>
    <rights>Music Follower</rights>
    <entry>
        <title type="html"><![CDATA[foo by bar]]></title>
        <id>1</id>
        <link href="http://example.com"/>
        <updated>1970-01-01T00:00:00.000Z</updated>
        <summary type="html"><![CDATA[<img src="http://example.com" />
          <h1>foo</h1><p>bar</p>
          <p>1970</p>]]></summary>
    </entry>
</feed>'
`;

snapshot[`global > GET / > should send body as "application/json" if accepts is "application/json" 1`] = `
'{
    "version": "https://jsonfeed.org/version/1",
    "title": "Music releases",
    "description": "Get the latest music releases from artist you follow",
    "items": [
        {
            "id": "1",
            "content_html": "<img src=\\\\"http://example.com\\\\" />\\\\n          <h1>foo</h1><p>bar</p>\\\\n          <p>1970</p>",
            "url": "http://example.com",
            "title": "foo by bar",
            "date_modified": "1970-01-01T00:00:00.000Z"
        }
    ]
}'
`;
